fluent-syntax/src/parser/mod.rs

```diff
@@ -1,14 +1,842 @@
+        vec![]
+    });
+
+    if pattern.is_none() && attributes.is_empty() {
+        return error!(
+            ErrorKind::ExpectedMessageField {
+                entry_id: id.name.to_string()
+            },
+            entry_start, ps.ptr
+        );
+    }
+
+    Ok(ast::Message {
+        id,
+        value: pattern,
+        attributes,
+        comment: None,
+    })
+}
+
+fn get_term<'p>(ps: &mut ParserStream<'p>, entry_start: usize) -> Result<ast::Term<'p>> {
+    ps.expect_byte(b'-')?;
+    let id = get_identifier(ps)?;
+    ps.skip_blank_inline();
+    ps.expect_byte(b'=')?;
+    ps.skip_blank_inline();
+
+    let value = get_value(ps)?;
+
+    ps.skip_blank_block();
+
+    let ptr = ps.ptr;
+    let attributes = get_attributes(ps).unwrap_or_else(|_| {
+        ps.ptr = ptr;
+        vec![]
+    });
+
+    if let Some(value) = value {
+        Ok(ast::Term {
+            id,
+            value,
+            attributes,
+            comment: None,
+        })
+    } else {
+        error!(
+            ErrorKind::ExpectedTermField {
+                entry_id: id.name.to_string()
+            },
+            entry_start, ps.ptr
+        )
+    }
+}
+
+fn get_value<'p>(ps: &mut ParserStream<'p>) -> Result<Option<ast::Value<'p>>> {
+    if ps.skip_to_value_start().is_none() {
+        return Ok(None);
+    }
+
+    if ps.is_current_byte(b'{') {
+        let start = ps.ptr;
+        ps.ptr += 1;
+        ps.skip_blank();
+        if ps.is_current_byte(b'*') || ps.is_current_byte(b'[') {
+            let variants = get_variants(ps)?;
+            ps.expect_byte(b'}')?;
+            return Ok(Some(ast::Value::VariantList { variants }));
+        }
+        ps.ptr = start;
+    }
+
+    let pattern = get_pattern(ps)?;
+
+    Ok(pattern.map(ast::Value::Pattern))
+}
+
+fn get_attributes<'p>(ps: &mut ParserStream<'p>) -> Result<Vec<ast::Attribute<'p>>> {
+    let mut attributes = vec![];
+
+    loop {
+        let line_start = ps.ptr;
+        ps.skip_blank_inline();
+        if !ps.is_current_byte(b'.') {
+            ps.ptr = line_start;
+            break;
+        }
+
+        match get_attribute(ps) {
+            Ok(attr) => attributes.push(attr),
+            Err(_) => {
+                ps.ptr = line_start;
+                break;
+            }
+        }
+    }
+    Ok(attributes)
+}
+
+fn get_attribute<'p>(ps: &mut ParserStream<'p>) -> Result<ast::Attribute<'p>> {
+    ps.expect_byte(b'.')?;
+    let id = get_identifier(ps)?;
+    ps.skip_blank_inline();
+    ps.expect_byte(b'=')?;
+    let pattern = get_pattern(ps)?;
+
+    match pattern {
+        Some(pattern) => Ok(ast::Attribute { id, value: pattern }),
+        None => error!(ErrorKind::MissingValue, ps.ptr),
+    }
+}
+
+fn get_identifier<'p>(ps: &mut ParserStream<'p>) -> Result<ast::Identifier<'p>> {
+    let mut ptr = ps.ptr;
+
+    while let Some(b) = ps.source.get(ptr) {
+        if ps.is_byte_alphabetic(*b) {
+            ptr += 1;
+        } else if ptr == ps.ptr {
```

Comment:
Yes. Since `get_identifier` is such a hot loop I tested tons of variants and this is the fastest one. I'm going to switch to a more readable one for the sake of landing it sooner, and we can reevaluate perf optimizations later.